streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/GraphTest.java

```diff
@@ -0,0 +1,414 @@
+            exception.getMessage());
+    }
+
+    @Test
+    public void testDisconnectedGraph() {
+        final Graph<Integer> graph1 = new Graph<>();
+        graph1.addEdge(0, 1, 1, 1, 1);
+        graph1.addEdge(2, 3, 2, 1, 2);
+        graph1.setSourceNode(0);
+        graph1.setSinkNode(1);
+        final Exception exception = assertThrows(IllegalStateException.class, graph1::solveMinCostFlow);
+        assertEquals("Input flow for node 3 is 2 which doesn't match output flow null",
+            exception.getMessage());
+    }
+
+    @Test
+    public void testDisconnectedGraphCrossSourceSink() {
+        final Graph<Integer> graph1 = new Graph<>();
+        graph1.addEdge(0, 1, 1, 1, 1);
+        graph1.addEdge(2, 3, 2, 1, 2);
+        graph1.setSourceNode(0);
+        graph1.setSinkNode(3);
+        final Exception exception = assertThrows(IllegalStateException.class, graph1::solveMinCostFlow);
+        assertEquals("Input flow for node 1 is 1 which doesn't match output flow null",
+            exception.getMessage());
+    }
+
+    @Test
+    public void testJustSourceSink() {
+        final Graph<Integer> graph1 = new Graph<>();
+        graph1.addEdge(0, 1, 1, 1, 1);
+        graph1.setSourceNode(0);
+        graph1.setSinkNode(1);
+        graph1.solveMinCostFlow();
+        assertEquals(1, graph1.totalCost());
+    }
+
+    @Test
+    public void testMinCostFlow() {
+        // Original graph, flow from 0 to 1 and 2 to 3
+        Map<Integer, Graph<Integer>.Edge> edges = graph.edges(0);
+        Graph<Integer>.Edge edge = edges.get(1);
+        assertEquals(1, edge.flow);
+        assertEquals(0, edge.residualFlow);
+
+        edge = edges.get(3);
+        assertEquals(0, edge.flow);
+        assertEquals(1, edge.residualFlow);
+
+        edges = graph.edges(2);
+        edge = edges.get(3);
+        assertEquals(1, edge.flow);
+        assertEquals(0, edge.residualFlow);
+
+        edge = edges.get(1);
+        assertEquals(0, edge.flow);
+        assertEquals(1, edge.residualFlow);
+
+        assertEquals(5, graph.totalCost());
+
+        graph.solveMinCostFlow();
+
+        assertEquals(2, graph.totalCost());
+
+        edges = graph.edges(0);
+        assertEquals(2, edges.size());
+
+        // No flow from 0 to 1
+        edge = edges.get(1);
+        assertEquals(0, edge.flow);
+        assertEquals(1, edge.residualFlow);
+
+        // Flow from 0 to 3 now
+        edge = edges.get(3);
+        assertEquals(1, edge.flow);
+        assertEquals(0, edge.residualFlow);
+
+        edges = graph.edges(2);
+        assertEquals(2, edges.size());
+
+        // No flow from 2 to 3
+        edge = edges.get(3);
+        assertEquals(0, edge.flow);
+        assertEquals(1, edge.residualFlow);
+
+        // Flow from 2 to 1 now
+        edge = edges.get(1);
+        assertEquals(1, edge.flow);
+        assertEquals(0, edge.residualFlow);
+    }
```

Comment:
Seems we do not check all conditions? Flow from 0->1 is shifted to 0->3, right?

Reply:
0->3 flow is checked on line 269?